strings/guid/guid.go

```diff
@@ -0,0 +1,46 @@
+// guid.go
+// description: Generate random globally unique identifiers (GUIDs).
+// details:
+// A GUID (globally unique identifier) is a 128-bit text string that
+// represents an identification (ID). Organizations generate GUIDs when
+// a unique reference number is needed to identify information on
+// a computer or network. A GUID can be used to ID hardware, software,
+// accounts, documents and other items. The term is also often used in
+// software created by Microsoft.
+// See more information on: https://en.wikipedia.org/wiki/Universally_unique_identifier
+// author(s) [cheatsnake](https://github.com/cheatsnake)
+// see guid_test.go
+
+// Package guid provides facilities for generating random globally unique identifiers.
+package guid
+
+import (
+	"crypto/rand"
+	"fmt"
+	"math/big"
+	"strings"
+)
+
+const pattern string = "xxxxxxxx-xxxx-4xxx-xxxx-xxxxxxxxxxxx"
+
+// New returns a randomly generated global unique identifier.
+func New() (string, error) {
+	var guid strings.Builder
+
+	for i := range pattern {
+		char := string(pattern[i])
+
+		if char == "x" {
+			random, err := rand.Int(rand.Reader, big.NewInt(16))
+			if err != nil {
+				return "", err
+			}
+
+			char = fmt.Sprintf("%x", random.Int64())
+		}
+
+		guid.WriteString(char)
+	}
```

Comment:
The following are non-blocking suggestions.

```suggestion
	for i, ch := range pattern {
		if i == 14 {
			guid.WriteRune(ch)
			continue
		}
		if ch == '-' {
			guid.WriteRune(ch)
			continue
		}
		random, err := rand.Int(rand.Reader, big.NewInt(16))
		if err != nil {
			return "", err
		}
		guid.WriteString(fmt.Sprintf("%x", random.Int64()))
	}
```

This adds a few lines but makes it somewhat easy to notice what your main intent of the function is. If you notice, there were two cases which were previously hidden from the previous implementation (case when `ch` was in index 14 or when `ch` was `-`), making it somewhat difficult if it turns out there was a problem. Having the problems conditions written as guard statements are much easier to wrap your head around. (Not something that is always possible but should be preferred over simply hiding edge cases)

Reply:
In that case, it may be a little unclear where the number 14 came from. Perhaps you should put the version index in a separate constant to avoid "magic numbers" in the code.

```suggestion
const pattern string = "xxxxxxxx-xxxx-4xxx-xxxx-xxxxxxxxxxxx"
const versionIndex int = 14

// New returns a randomly generated global unique identifier.
func New() (string, error) {
	var guid strings.Builder

	for i, ch := range pattern {
	        if i == versionIndex {
		        guid.WriteRune(ch)
		        continue
	        }
	        if ch == '-' {
		        guid.WriteRune(ch)
		        continue
	        }
	        random, err := rand.Int(rand.Reader, big.NewInt(16))
	        if err != nil {
		        return "", err
	        }
	        guid.WriteString(fmt.Sprintf("%x", random.Int64()))
        }
	return guid.String(), nil
}
```